### PR TITLE
adding support for JSON params

### LINKED ISSUE #145: Epi Crashes on missing nested value object parameter
We discovered that some users have added sql parameters referencing nested values such as `--@color varchar pants.color`

If `pants` doesn't exist, the epi worker dies. 

Let's fix this so it handles it gracefully.

### PR DESCRIPTION
**Pull request by B0R1K**
We have a need for applications to be able to pass objects to a SQL template and have the object serialized into an escaped string. 
We are achieving this by introducing a faux data type of "json". 

With this update, an epiquery template can write a parameter like: 
```
--@myJson json jsonInput
```
Any object with the key of "jsonInput" in this example will be converted to a serialized Json string that can be ingested in TSQL using OPENJSON and deserialized. 

Any json data type will use "un-escaped" context. Prior to this change, context with strings was run through a function and we performed special character replacements on it. A workaround had been added in the sanitize step where if a parameter's name began with "json...", the special character replacement would be skipped. For our new json data type, the parameter can be named anything and the value will always be un-escaped. 

## Nerdy Details
* Fixed #145: _.reduce was hitting a condition when a nested parameter wasn't passed, it attempted to find an attribute of undefined and crashed the epi worker. I added a test for this bug as well as a positive test case to cover the expected case. 
* `unEscaptedContext` was being added to the context object in `sanitizeInput`, but it was in the wrong place and thus not actually unescaped. It was the same as the escapedContext. It was only used in one place, which I was modifying anyway. You'll notice that I moved it in SanitizeInput, but use it differently in mssql_base. 
* I made some small performance improvements to request.coffee's SanitizeInput method. We are bypassing the headers in the context and if we find a key starting with "json", we get out BEFORE we loop through all the special chars rather than continually test inside the loop

### CODE REVIEW

File: difftest/templates/test/param_json_is_json.sql [hunk 1]
```diff
@@ -0,0 +1,14 @@
+--parameters:
+--@myJson json myJson
+
+
+IF ISJSON(@myJson) > 0
+    SELECT 'Valid Json' AS 'Result';
+
+declare @apples table (type varchar(32))
```
Comment (jimclouse) [inline difftest/templates/test/param_json_is_json.sql:8]: lets remove. unused

File: difftest/results/param_json_large_json_data [hunk 2]
```diff
@@ -0,0 +1,14 @@
+{
```
[the remark this answers is not in the document]
Comment (jimclouse) [reply difftest/results/param_json_large_json_data]: removed the file

File: src/drivers/mssql_base.coffee [hunk 3]
```diff
@@ -87,18 +90,23 @@ class MSSQLDriver extends events.EventEmitter
...
-      # duplicating 's
-      value = _.reduce value.split('.'), (doc,prop) ->
-        doc[prop]
-      , context.unEscapedTemplateContext
+      # If the variable type is json, use the unescapted context
+      # this is done here because the sanitize input step isn't aware of types
+      # and would have to rely on the parameter name and hokey tree traversal depth monitoring
+      contextToUse = context.templateContext
+      if type == "json"
+        contextToUse = context.unEscapedTemplateContext
```
Comment (JamesGally) [inline src/drivers/mssql_base.coffee:102]: @jimclouse I'd like to talk about this code later today, also the comment has a typo (hokey)

**Review by igroff — APPROVED** [review]
I did add a comment, but it was just that: a comment. Just there to help us remember where we came from.

File: src/request.coffee [hunk 4]
```diff
@@ -204,17 +204,21 @@ collectStats = (context, callback) ->
...
+      return if key.toLowerCase and key.toLowerCase().startsWith 'json'  
       _.each Object.keys(special_characters), (keyCode) ->
-        # do not escape our JSON data since it's JSON and does it's own thing
-        # oh, and don't try to lower case things that don't have the toLowerCase
-        # method, such as numbers which are the 'keys' of arrays
-        return if key.toLowerCase and key.toLowerCase().startsWith 'json'
         def = special_characters[keyCode]
         value = value.replace def.regex, def.replace
       parent[key] = value
-  context.unEscapedTemplateContext = _.cloneDeep context.templateContext
```
[the remark this answers is not in the document]
Comment (jimclouse) [reply src/request.coffee]: @igroff  this was super confusing, as I re-purposed this variable. However, what I missed is that we have 2 notions of "unescaped" now, which we need to maintain for backwards compatibility. My latest updates resolve this strangeness in the variable names. 

The upshot is that we have one replacement of "special characters" where we replace Microsoft quotes and things, and then a second replacement to strip out single quotes. Our newest version leaves "unEscaptedTemplateContext" alone and introduces "originalTemplateContext". 

* unEscaptedTemplateContext: Has special characters stripped (unless parameter name begins with "json") but keeps all single quotes. This is used only for parameterized sql, not mustache. 
* originalTemplateContext: Has nothing replaced. keeps all special characters and all single quotes. (only used for "json" parameter types)

**Comment by jimclouse** [comment]
Closing this PR and re-opening a new one with a dash-separated branch name for GDS. 
new PR: https://github.com/glg/epiquery2/pull/148

**Pull request closed without merging**